README.md
# ACF Field Selector Field

A field for Advanced Custom Fields which allows users to select a list of created custom fields

-----------------------

### Description

The filed selector field allows the selection of other custom fields. This is useful if you want to build a form from the custom fields a user has selected for example. When creating the field you have the following options:

* Include/Exclude Field Types
* Include/Exclude Field Groups

### Compatibility

This ACF field type is compatible with:

* ACF 5
* ACF 4


### Installation

1. Copy the `acf-field_selector_field` folder into your `wp-content/plugins` folder
2. Activate the Advanced Custom Fields: Field Selector plugin via the plugins admin page
3. Create a new field via ACF and select the Field Selector type
4. Please refer to the description for more info regarding the field type settings

### Usage

For developers I've included a filter which allows you to further filter selected fields. At the moment this filter is used to make sure that included and excluded types and groups are reflected in the selectable list.

```php
add_filter( 'acffsf/item_filters', 'selectable_item_filter', 10, 2 )
```

The first parameter is the list of items to modify, the second is the setting field.


### Changelog

= 3.0.2 =
* Compatibility with 4.1

= 3.0.1 =
* Made sure search is case-insensitive
* Made sure value was json decoded when returned to the_field

= 3.0 =
* Complete rewrite with custom controls

= 2.0 =
* Added ACF 5 Support
* Removed ACF 3 Support

= 1.0 =
* Initial Release.

-----------------------

* Readme : https://github.com/danielpataki/acf-google_font_selector/blob/master/acf-google_font_selector/trunk/readme.txt
* WordPress repository: https://wordpress.org/plugins/acf-field-selector-field/

-----------------------

Special thanks to [Elliot Condon](http://elliotcondon.com) for making the wonderful [ACF plugin](advancedcustomfields.com) in the first place.


ChangeLog.md
# 3.0.2
* Compatibility with 4.1

# 3.0.1
- Made sure search is case-insensitive
- Made sure value was json decoded when returned to the_field

# 3.0
- Complete rewrite with custom controls

# 2.0
- Added ACF 5 Support
- Removed ACF 3 Support

# 1.0
- Initial Release.


field_selector-v3.php
<?php

class acf_field_field_selector extends acf_Field
{

	// vars
	var $settings,
		$defaults;


	/*--------------------------------------------------------------------------------------
	*
	*	Constructor
	*	- This function is called when the field class is initalized on each page.
	*	- Here you can add filters / actions and setup any other functionality for your field
	*
	*	@author Elliot Condon
	*	@since 2.2.0
	*
	*-------------------------------------------------------------------------------------*/

	function __construct($parent)
	{

		// do not delete!
		parent::__construct($parent);

		// vars
		$this->name = 'field_selector';
		$this->label = __('Field Selector', 'acf');
		$this->category = __("Relational",'acf'); // Basic, Content, Choice, etc
		$this->defaults = array(
			'allowed_groups'     => '',
			'allowed_types'      => '',
			'exclude_types'      => '',
			'max'                => '',
			'return_value'       => 'key',
			'field_type'         => 'autocomplete',
			'additional_fields'  => '',
		);


		// do not delete!
		parent::__construct();

		$this->l10n = array(
			'max'		=> __("Maximum fields reached ( {max} fields )",'acf'),
			'tmpl_li'	=> '
				<li>
					<a href="#" data-name="<%= name %>" data-value="<%= value %>"><%= title %><span class="acf-button-remove"></span></a>
					<input type="hidden" name="<%= name %>[]" value="<%= value %>" />
				</li>
			'
		);


		// settings
		$this->settings = array(
			'path' => apply_filters('acf/helpers/get_path', __FILE__),
			'dir' => apply_filters('acf/helpers/get_dir', __FILE__),
			'version' => '1.0.0'
		);

  }


	/*
	*  Field Groups Array
	*
	*  Generates an array of all field groups
	*
	*/
	function get_field_group_array() {
		$field_groups = get_posts( array( 'post_type' => 'acf', 'posts_per_page' => -1 ) );
		$groups = array();

		foreach( $field_groups as $group ) {
			$groups[$group->ID] = $group->post_title;
		}

		return $groups;
	}



 	/*
  *  helpers_get_path
  *
  *  @description: calculates the path (works for plugin / theme folders)
  *  @since: 3.6
  *  @created: 30/01/13
  */

  function helpers_get_path($file)
  {
    return trailingslashit(dirname($file));
  }


  /*
  *  helpers_get_dir
  *
  *  @description: calculates the directory (works for plugin / theme folders)
  *  @since: 3.6
  *  @created: 30/01/13
  */

  function helpers_get_dir($file)
  {
    $dir = trailingslashit(dirname($file));
    $count = 0;


    // sanitize for Win32 installs
    $dir = str_replace('\\', '/', $dir);


    // if file is in plugins folder
    $wp_plugin_dir = str_replace('\\', '/', WP_PLUGIN_DIR);
    $dir = str_replace($wp_plugin_dir, WP_PLUGIN_URL, $dir, $count);


    if($count < 1)
    {
      // if file is in wp-content folder
      $wp_content_dir = str_replace('\\', '/', WP_CONTENT_DIR);
      $dir = str_replace($wp_content_dir, WP_CONTENT_URL, $dir, $count);
    }


    if($count < 1)
    {
      // if file is in ??? folder
      $wp_dir = str_replace('\\', '/', ABSPATH);
      $dir = str_replace($wp_dir, site_url('/'), $dir);
    }

    return $dir;
  }


	/*--------------------------------------------------------------------------------------
	*
	*	create_options
	*	- this function is called from core/field_meta_box.php to create extra options
	*	for your field
	*
	*	@params
	*	- $key (int) - the $_POST obejct key required to save the options to the field
	*	- $field (array) - the field object
	*
	*	@author Elliot Condon
	*	@since 2.2.0
	*
	*-------------------------------------------------------------------------------------*/

	function create_options($key, $field)
	{
		$field = array_merge($this->defaults, $field);
		$key = $field['name'];


		// Create Field Options HTML
		?>

		<tr class="field_option field_option_<?php echo $this->name; ?>">
			<td class="label">
				<label><?php _e("Display Type",'acf'); ?></label>
			</td>
			<td>
				<?php
				do_action('acf/create_field', array(
					'type'	=>	'select',
					'name'	=>	'fields['.$key.'][field_type]',
					'value'	=>	$field['field_type'],
					'choices' => array(
						__( 'Multiple Values', 'acf' ) => array(
							'autocomplete' => __( 'Autocomplete', 'acf' ),
							'checkbox' => __( 'Checkbox', 'acf' ),
							'multi_select' => __( 'Multi Select', 'acf' )
						),
						__( 'Single Value', 'acf' ) => array(
							'radio' => __( 'Radio Buttons', 'acf' ),
							'select' => __( 'Select', 'acf' )
						)
					)
				));
				?>
			</td>
		</tr>



		<tr class="field_option field_option_<?php echo $this->name; ?>">
			<td class="label">
				<label><?php _e("Return Value",'acf'); ?></label>
			</td>
			<td>
				<?php
				do_action('acf/create_field', array(
					'type'	=>	'radio',
					'name'	=>	'fields['.$key.'][return_value]',
					'value'	=>	$field['return_value'],
					'choices' => array(
						'key' => __( 'Field Key', 'acf' ),
						'name' => __( 'Field Name', 'acf' ),
						'object' => __( 'Field Object', 'acf' ),
					)
				));
				?>
			</td>
		</tr>


		<tr class="field_option field_option_<?php echo $this->name; ?>">
			<td class="label">
				<label><?php _e("Maximum items",'acf'); ?></label>
			</td>
			<td>
				<?php
				do_action('acf/create_field', array(
					'type'	=>	'number',
					'name'	=>	'fields['.$key.'][max]',
					'value'	=>	$field['max'],
				));
				?>
			</td>
		</tr>

		<tr class="field_option field_option_<?php echo $this->name; ?>">
			<td class="label">
				<label><?php _e("Allowed Field Types",'acf'); ?></label>
				<p class="description"><?php _e( "Leave empty to allow all, otherwise one type per row", 'acf' ) ?></p>
			</td>
			<td>
				<?php
				do_action('acf/create_field', array(
					'type'	=>	'textarea',
					'name'	=>	'fields['.$key.'][allowed_types]',
					'value'	=>	implode( "\n", $field['allowed_types'] ),
				));
				?>
			</td>
		</tr>



		<tr class="field_option field_option_<?php echo $this->name; ?>">
			<td class="label">
				<label><?php _e("Excluded Field Types",'acf'); ?></label>
				<p class="description"><?php _e( "Set field types to exclude specifically, one per row", 'acf' ) ?></p>
			</td>
			<td>
				<?php
				do_action('acf/create_field', array(
					'type'	=>	'textarea',
					'name'	=>	'fields['.$key.'][exclude_types]',
					'value'	=>	implode( "\n", $field['exclude_types'] ),
				));
				?>
			</td>
		</tr>

		<tr class="field_option field_option_<?php echo $this->name; ?>">
			<td class="label">
				<label><?php _e("Allowed Field Groups",'acf'); ?></label>
				<p class="description"><?php _e( "Leave empty to allow all, otherwise one type per row", 'acf' ) ?></p>
			</td>
			<td>
				<?php
				do_action('acf/create_field', array(
					'type'	=>	'select',
					'multiple' => true,
					'name'	=>	'fields['.$key.'][allowed_groups]',
					'value'	=>	$field['allowed_groups'],
					'choices' => $this->get_field_group_array()
				));
				?>
			</td>
		</tr>
		<?php
	}



	/*--------------------------------------------------------------------------------------
	*
	*	create_field
	*	- this function is called on edit screens to produce the html for this field
	*
	*	@author Elliot Condon
	*	@since 2.2.0
	*
	*-------------------------------------------------------------------------------------*/

	function create_field($field)
	{
		$args = array(
			'post_type' => 'acf',
			'posts_per_page' => -1
		);

		if( !empty( $field['allowed_groups'] ) ) {
			$args['include'] = $field['allowed_groups'];
		}

		$field_groups = get_posts( $args );

		$fields = array();
		if( !empty( $field_groups ) ) {
			foreach( $field_groups as $field_group ) {
				$custom_fields = get_post_meta( $field_group->ID );
				foreach( $custom_fields as $key => $customfield ) {
					if( substr_count( $key, 'field_' ) > 0 ) {
						$customfield = get_post_meta( $field_group->ID, $key, true );
						$fields[$customfield['label']] = array(
							'group' => $field_group->post_title,
							'field' => $customfield
						);
					}
				}
			}
			ksort($fields);
		}


		foreach( $fields as $name => $data ) {
			if( !empty( $field['allowed_types'] ) && !in_array( $data['field']['type'], $field['allowed_types'] ) ) {
				unset( $fields[$name] );
			}
			if( !empty( $field['exclude_types'] ) && in_array( $data['field']['type'], $field['exclude_types'] ) ) {
				unset( $fields[$name] );
			}
		}


		?>
		<div>

		<?php

			$multiselect = ( $field['field_type'] == 'multi_select' ) ? 'multiple="multiple"' : '';

			$fields_by_group = array();
			foreach( $fields as $name => $data ) {
				$fields_by_group[$data['group']][] = $data['field'];
			}

			$fields_by_key = array();
			foreach( $fields as $name => $data ) {
				$fields_by_key[$data['field']['key']]['field'] = $data['field'];
				$fields_by_key[$data['field']['key']]['group'] = $data['group'];
			}


			switch ( $field['field_type'] ) {
				case 'select' :
				case 'multi_select' :
					echo "<select " . $multiselect . " name='" . $field['name'] . "[]'>";
					foreach( $fields_by_group as $name => $group ) {
						echo '<optgroup label="' . $name . '">';
						foreach( $group as $item ) {
							echo '<option ' . selected( $field['value'], $item['key'] , false ) . ' value="' . $item['key'] . '">' . $item['label'] . '</value>';
						}
						echo '</optgroup>';
					}
					echo '</select>';
					break;
				case 'radio' :
					echo '<ul class="acf-radio-list radio vertical acffs-option-group-list">';
					foreach( $fields_by_group as $name => $group ) {
						echo '<li><span class="acffs-option-group-label">' . $name . '</span><ul>';
						foreach( $group as $item ) {
							echo '<li><label><input ' . checked( $field['value'], $item['key'] , false ) . ' type="radio" name="' . $field['name'] . '[]" value="' . $item['key'] . '">' . $item['label'] . '</label></li>';
						}
						echo '</ul></li>';
					}
					echo '</ul>';
					break;
				case 'checkbox' :
					echo '<ul class="acf-checkbox-list checkbox vertical acffs-option-group-list">';
					foreach( $fields_by_group as $name => $group ) {
						echo '<li><span class="acffs-option-group-label">' . $name . '</span><ul>';
						foreach( $group as $item ) {
							echo '<li><label><input ' . checked( $field['value'], $item['key'] , false ) . ' type="checkbox" name="' . $field['name'] . '[]" value="' . $item['key'] . '">' . $item['label'] . '</label></li>';
						}
						echo '</ul></li>';
					}
					echo '</ul>';
					break;
				case 'autocomplete' :
				default :

					$attributes = array(
						'max' => $field['max'],
						's' => '',
						'field_key' => $field['key']
					);

				?>
				<div class="acffs-autocomplete-container" <?php foreach( $attributes as $k => $v ): ?> data-<?php echo $k; ?>="<?php echo $v; ?>"<?php endforeach; ?>>
					<input type="hidden" name="<?php echo $field['name']; ?>" value="" />

				<!-- Left List -->
				<div class="acffs-autocomplete-left">
					<table class="widefat">
						<thead>
							<tr>
								<th>
									<input class="acffs-autocomplete-search" placeholder="<?php _e("Search...",'acf'); ?>" type="text" id="custom_field_selector_<?php echo $field['name']; ?>" />
								</th>
							</tr>
						</thead>
					</table>
					<ul class="bl acffs-autocomplete-list">
						<?php
							if( !empty( $fields ) ) :
								foreach( $fields as $customfield ) :
								$hidden = ( !empty( $field['value'] ) && is_array( $field['value'] ) && in_array( $customfield['field']['key'], $field['value'] ) ) ? 'class="hide"' : '';
							?>
						<li <?php echo $hidden ?>>
							<a href="#" data-name="<?php echo $customfield['field']['label'] ?> <?php echo $customfield['group'] ?>" data-value="<?php echo $customfield['field']['key'] ?>"><?php echo $customfield['field']['label'] ?> <span class='additional-data'><?php echo $customfield['group'] ?></span> <span class="acf-button-add"></span></a>
						</li>
						<?php endforeach; endif ?>

					</ul>
				</div>
				<!-- /Left List -->

				<!-- Right List -->
				<div class="acffs-autocomplete-right">
					<ul class="bl acffs-autocomplete-list">
					<?php
					if( !empty( $field['value'] ) && is_array( $field['value'] ) )
					{
						foreach( $field['value'] as $value )
						{

							if( !empty( $fields_by_key[$value] ) ) {

							$customfield = $fields_by_key[$value];
							echo '<li>
								<a href="#" class="" data-name="' . $customfield['field']['label'] . '" data-value="' . $customfield['field']['key'] . '">' . $customfield['field']['label'] . '<span class="additional-data">' . $customfield['group'] .'</span> <span class="acf-button-remove"></span></a>
								<input type="hidden" name="' . $field['name'] . '[]" value="' . $customfield['field']['key'] . '" />
							</li>';

							}


						}
					}

					?>
					</ul>
				</div>

				</div>
				<!-- / Right List -->


				<?php


			}
		?>

		</div>
		<?php
	}


	/*--------------------------------------------------------------------------------------
	*
	*	admin_print_scripts / admin_print_styles
	*	- this function is called in the admin_print_scripts / admin_print_styles where
	*	your field is created. Use this function to register css and javascript to assist
	*	your create_field() function.
	*
	*	@author Elliot Condon
	*	@since 3.0.0
	*
	*-------------------------------------------------------------------------------------*/

	function admin_print_scripts()
	{
		// register acf scripts
		wp_register_script('acf-input-field_selector', $this->settings['dir'] . 'js/input.js', array('acf-input'), $this->settings['version']);


		// scripts
		wp_enqueue_script(array(
			'acf-input-field_selector',
		));

	}

	function admin_print_styles()
	{
		// Note: This function can be removed if not used
		wp_register_style('acf-input-field_selector', $this->settings['dir'] . 'css/input.css', array('acf-input'), $this->settings['version']);
		// styles
		wp_enqueue_style(array(
			'acf-input-field_selector',
		));

	}



	/*--------------------------------------------------------------------------------------
	*
	*	get_value_for_api
	*	- called from your template file when using the API functions (get_field, etc).
	*	This function is useful if your field needs to format the returned value
	*
	*	@params
	*	- $post_id (int) - the post ID which your value is attached to
	*	- $field (array) - the field object.
	*
	*	@author Elliot Condon
	*	@since 3.0.0
	*
	*-------------------------------------------------------------------------------------*/

	function get_value_for_api($post_id, $field)
	{
		$value = get_field( $field['key'], $post_id  );
		if( !empty( $value ) ) {
			if( $field['return_value'] == 'object' ) {
				foreach( $value as $key => $item ) {
					$value[$key] = get_field_object( $item );
				}
			}

			if( $field['return_value'] == 'name' ) {
				foreach( $value as $key => $item ) {
					$field_object = get_field_object( $item );
					$value[$key] = $field_object['name'];
				}
			}
		}
		

		return $value;

	}

	function update_value($post_id, $field, $value)
	{
		$field['allowed_types'] = explode( "\n", $field['allowed_types'] );
		$field['allowed_types'] = array_map( 'trim', $field['allowed_types'] );
		$field['allowed_types'] = array_filter( $field['allowed_types'] );

		$field['exclude_types'] = explode( "\n", $field['exclude_types'] );
		$field['exclude_types'] = array_map( 'trim', $field['exclude_types'] );
		$field['exclude_types'] = array_filter( $field['exclude_types'] );

		return $field;
	}

}

?>

acf-field_selector.php
<?php
/*
Plugin Name: Advanced Custom Fields: Field Selector
Plugin URI: http://github.com
Description: A field which allows you to select other custom fields
Version: 1.0.0
Author: Daniel Pataki
Author URI: http://danielpataki.com
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html
*/


class acf_field_field_selector_plugin
{
	/*
	*  Construct
	*
	*  @description:
	*  @since: 3.6
	*  @created: 1/04/13
	*/

	function __construct()
	{
		// set text domain
		/*
		$domain = 'acf-field_selector';
		$mofile = trailingslashit(dirname(__File__)) . 'lang/' . $domain . '-' . get_locale() . '.mo';
		load_textdomain( $domain, $mofile );
		*/


		// version 4+
		add_action('acf/register_fields', array($this, 'register_fields'));


		// version 3-
		add_action('init', array( $this, 'init' ), 5);
	}


	/*
	*  Init
	*
	*  @description:
	*  @since: 3.6
	*  @created: 1/04/13
	*/

	function init()
	{
		if(function_exists('register_field'))
		{
			register_field('acf_field_field_selector', dirname(__File__) . '/field_selector-v3.php');
		}
	}

	/*
	*  register_fields
	*
	*  @description:
	*  @since: 3.6
	*  @created: 1/04/13
	*/

	function register_fields()
	{
		include_once('field_selector-v4.php');
	}

}

new acf_field_field_selector_plugin();

?>

acf-field_selector_field/trunk/js/input.js
(function($){

	function initialize_field( $el ) {

		//$el.doStuff();

	}

	function add_to_selected_list( key, $field ) {
		var current = $field.val();
		var values = [];

		if( current === '' || typeof current === 'undefined' || current === null || current === '{}' ) {
			values.push( key );
		}
		else {
			var values = $.parseJSON( current );
			values.push( key )
		}

		values = JSON.stringify( values );
		$field.val( values );

	}

	function sort_selected( $multiselect ) {
		var $selected = $multiselect.find( '.selected.field li' );
		var $input_field = $multiselect.find( '#field-value' );

		values = [];
		if( $selected.length > 0 ) {
			$.each( $selected, function() {
				values.push( $(this).data('key') );
			})
		}

		values = JSON.stringify( values );
		$input_field.val( values );

	}

	function add_to_list_ui( $item, $list, $other_list ) {
		$item.appendTo( $list );
	}

	function order_list( $list ) {
		var list_array = [];
		$.each( $list.find('li'), function() {
			list_array.push( $(this).data('search_term') );
		})
		list_array.sort();

		$.each( list_array, function() {
			$list.append( $list.find("li[data-search_term='" + this + "']") );
		})

	}

	function remove_from_selected_list( key, $field ) {
		var current = $.parseJSON( $field.val() );

		current = $.grep( current, function(value) {
		  return value != key;
		});

		values = JSON.stringify( current );
		if( values === '[]' ) {
			values = '';
		}
		$field.val( values );

	}

	function get_field_search_elements( query, $selectable ) {
		var toShow = $();
		$.each( $selectable.find('li'), function() {
			if( $(this).data('search_term').indexOf( query ) > -1 ) {
				var $element = $( 'li[data-key="' + $(this).data('key') + '"]' );
				toShow = toShow.add( $element );
			}
		});

		return toShow;
	}


	if( typeof acf.add_action !== 'undefined' ) {

		/*
		*  ready append (ACF5)
		*
		*  These are 2 events which are fired during the page load
		*  ready = on page load similar to $(document).ready()
		*  append = on new DOM elements appended via repeater field
		*
		*  @type	event
		*  @date	20/07/13
		*
		*  @param	$el (jQuery selection) the jQuery element which contains the ACF fields
		*  @return	n/a
		*/

		acf.add_action('ready append', function( $el ){

			// search $el for fields of type 'FIELD_NAME'
			acf.get_fields({ type : 'field_selector'}, $el).each(function(){

				initialize_field( $(this) );



			});

		});


	} else {


		/*
		*  acf/setup_fields (ACF4)
		*
		*  This event is triggered when ACF adds any new elements to the DOM.
		*
		*  @type	function
		*  @since	1.0.0
		*  @date	01/01/12
		*
		*  @param	event		e: an event object. This can be ignored
		*  @param	Element		postbox: An element which contains the new HTML
		*
		*  @return	n/a
		*/

		$(document).live('acf/setup_fields', function(e, postbox){

			$(postbox).find('.field[data-field_type="field_selector"]').each(function(){

				initialize_field( $(this) );

				var $this = $(this);

    			$(this).find('.selected.field').sortable({
					axis: 'y',
					containerment: 'parent',
					cursor: 'move',
					items: 'li',
					helper: 'clone',
					placeholder: 'sortable-placeholder',
					forcePlaceholderSize: true,
					opacity: 0.5,
					revert: 100,
					stop: function() {
						sort_selected( $this );
					}

				});
    			$(this).find('.selected').disableSelection();


				$( document ).on( 'click', '.multiselector .selectable li', function() {
					var $multiselector = $(this).parents( '.multiselector' );
					var $selected = $multiselector.find( '.selected ul' );
					var $selectable = $multiselector.find( '.selectable ul' );
					var $input_field = $multiselector.find( '#field-value' );
					var key = $(this).data('key');

					add_to_list_ui( $(this), $selected, $selectable );
					add_to_selected_list( key, $input_field );
				})

				$( document ).on( 'click', '.multiselector .selected li', function() {
					var $multiselector = $(this).parents( '.multiselector' );
					var $selectable = $multiselector.find( '.selectable ul' );
					var $selected = $multiselector.find( '.selected ul' );
					var $input_field = $multiselector.find( '#field-value' );
					var key = $(this).data('key');

					add_to_list_ui( $(this), $selectable, $selected );
					order_list( $selectable );
					remove_from_selected_list( key, $input_field );
				})

				$( document ).on( 'keyup', '.multiselector #field-search', function() {
					var query = $(this).val();
					var $multiselector = $(this).parents( '.multiselector' );
					var $selectable = $multiselector.find( '.selectable ul' );

					var $search_elements = get_field_search_elements( query, $selectable );

					$selectable.find('li').hide();
					$search_elements.show();

				})


			});

		});


	}


})(jQuery);


acf-field_selector_field/trunk/acf-field_selector-v4.php
<?php

class acf_field_field_selector extends acf_field {

	var $settings,
		$defaults;


	/*
	*  __construct
	*
	*  Set name / label needed for actions / filters
	*
	*  @since	3.6
	*  @date	23/01/13
	*/

	function __construct()
	{

		$this->name = 'field_selector';
		$this->label = __('Field Selector', 'acf-field_selector');
		$this->category = __("Choice",'acf');
		$this->defaults = array(
			'group_filtering' => 'include',
			'groups' => '',
			'type_filtering'  => 'include',
			'types'  => ''
		);
		$this->common = new acf_field_field_selector_common();

    	parent::__construct();

		$this->settings = array(
			'path' => apply_filters('acf/helpers/get_path', __FILE__),
			'dir' => apply_filters('acf/helpers/get_dir', __FILE__),
			'version' => '1.0.0'
		);
		
		add_filter( 'acffsf/item_filters', array( 'acf_field_field_selector_common', 'type_filter' ), 10, 2 );
		add_filter( 'acffsf/item_filters', array( 'acf_field_field_selector_common', 'group_filter' ), 10, 2 );

	}


	/*
	*  create_options()
	*
	*  Create extra options for your field. This is rendered when editing a field.
	*  The value of $field['name'] can be used (like below) to save extra data to the $field
	*
	*  @type	action
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	$field	- an array holding all the field's data
	*/

	function create_options( $field )
	{

		$field = array_merge($this->defaults, $field);
		if( !empty( $field['types'] ) ) {
			$field['types'] = implode(',', $field['types']);
		}
		else {
			$field['types'] = '';
		}

		if( !empty( $field['groups'] ) ) {
			$field['groups'] = implode(',', $field['groups']);
		}
		else {
			$field['groups'] = '';
		}

		// key is needed in the field names to correctly save the data
		$key = $field['name'];


		// Create Field Options HTML
		?>
<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Group Filtering",'acf'); ?></label>
		<p class="description"><?php _e("Enter group id numbers separated by commas to include or exclude them.",'acf'); ?></p>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'		=>	'radio',
			'name'		=>	'fields['.$key.'][group_filtering]',
			'value'		=>	$field['group_filtering'],
			'layout'	=>	'horizontal',
			'choices'	=>	array(
				'include' => __('Include'),
				'exclude' => __('Exclude'),
			)
		));

		?>

		<?php

		do_action('acf/create_field', array(
			'type'		=>	'text',
			'name'		=>	'fields['.$key.'][groups]',
			'value'		=>	$field['groups']
		));

		?>


	</td>
</tr>


<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Type Filtering",'acf'); ?></label>
		<p class="description"><?php _e("Enter type slugs separated by commas to include or exclude them.",'acf'); ?></p>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'		=>	'radio',
			'name'		=>	'fields['.$key.'][type_filtering]',
			'value'		=>	$field['type_filtering'],
			'layout'	=>	'horizontal',
			'choices'	=>	array(
				'include' => __('Include'),
				'exclude' => __('Exclude'),
			)
		));

		?>

		<?php

		do_action('acf/create_field', array(
			'type'		=>	'text',
			'name'		=>	'fields['.$key.'][types]',
			'value'		=>	$field['types']
		));

		?>


	</td>
</tr>



		<?php
	}


	/*
	*  create_field()
	*
	*  Create the HTML interface for your field
	*
	*  @param	$field - an array holding all the field's data
	*
	*  @type	action
	*  @since	3.6
	*  @date	23/01/13
	*/

	function create_field( $field )
	{
		?>
		<div class='multiselector'>
			<div class='selectable-container field-container'>
				<div class='title'><?php _e( 'Available Fields', 'acf-field_selector' ) ?></div>
				<div class='search'><input type='text' id="field-search" placeholder='Type to search...'></div>
				<div class='selectable field'>
					<?php
					$selectable = $this->get_items( $field, $this->get_selectable_item_fields( $field ) );
					$this->common->show_items( $selectable );
					?>
				</div>
			</div>
			<div class='selected-container field-container'>
				<div class='title'><?php _e( 'Selected Fields', 'acf-field_selector' ) ?></div>
				<div class='message'><?php _e( 'Drag and drop to re-order your selection', 'acf-field_selector' ) ?></div>
				<div class='selected field'>
					<?php
						$selected = $this->get_items( $field, $this->get_selected_item_fields( $field ), false );
						$this->common->show_items( $selected );
					?>
				</div>
			</div>
			<input type='hidden' id='field-value' name='<?php echo esc_attr($field['name']) ?>' value="<?php echo esc_attr($field['value']) ?>">
		</div>
		<?php
	}


	/*
	*  input_admin_enqueue_scripts()
	*
	*  This action is called in the admin_enqueue_scripts action on the edit screen where your field is created.
	*  Use this action to add CSS + JavaScript to assist your create_field() action.
	*
	*  $info	http://codex.wordpress.org/Plugin_API/Action_Reference/admin_enqueue_scripts
	*  @type	action
	*  @since	3.6
	*  @date	23/01/13
	*/

	function input_admin_enqueue_scripts()
	{


		// register ACF scripts
		wp_register_script( 'acf-input-field_selector', $this->settings['dir'] . 'js/input.js', array('acf-input'), $this->settings['version'] );
		wp_register_style( 'acf-input-field_selector', $this->settings['dir'] . 'css/input.css', array('acf-input'), $this->settings['version'] );

		wp_enqueue_script( 'jquery-ui-sortable' );


		// scripts
		wp_enqueue_script(array(
			'acf-input-field_selector',
		));

		// styles
		wp_enqueue_style(array(
			'acf-input-field_selector',
		));


	}




	/*
	*  update_field()
	*
	*  This filter is applied to the $field before it is saved to the database
	*
	*  @type	filter
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	$field - the field array holding all the field options
	*  @param	$post_id - the field group ID (post_type = acf)
	*
	*  @return	$field - the modified field
	*/

	function update_field( $field, $post_id )
	{

		if( !empty( $field['types'] ) ) {
			$field['types'] = array_map( 'trim', explode( ',', $field['types'] ) );
		}
		else {
			$field['types'] = array();
		}

		if( !empty( $field['groups'] ) ) {
			$field['groups'] = array_map( 'trim', explode( ',', $field['groups'] ) );
		}
		else {
			$field['groups'] = array();
		}

		return $field;
	}


	function sort_items_by_label($a, $b) {
		return strcmp( $a["label"], $b["label"] );
	}

	function get_items( $field, $items, $sort = true ) {
		$final_items = array();
		if( !empty( $items ) ) {
			foreach( $items as $item ) {
				$item_value = unserialize( $item->meta_value );

				$item_value['group'] = array(
					'ID' => $item->post_id,
					'post_title' => get_the_title( $item->post_id )
				);

				$final_items[$item_value['key']] = $item_value;
			}

			if( $sort == true ) {
				usort($final_items, array( $this, 'sort_items_by_label' ) );
			}

		}

		$final_items = apply_filters( 'acffsf/item_filters', $final_items, $field, $final_items );

		return $final_items;

	}

	function get_selectable_item_fields( $field ) {
		global $wpdb;

		$field_keys = array();
		$field_keys_query = 9999;

		if( !empty( $field['value'] ) ) {
			$field_keys = json_decode( $field['value'], true );
			$field_keys_query = "'" . implode( "', '", $field_keys ) . "'";
		}

		$fields = $wpdb->get_results( "SELECT meta_key, meta_value, post_id FROM $wpdb->postmeta WHERE post_ID IN (SELECT ID FROM $wpdb->posts WHERE post_type = 'acf' AND post_status = 'publish' ) AND meta_key LIKE 'field_%'
		AND meta_key NOT IN ( $field_keys_query )  " );
		return $fields;

	}

	function get_selected_item_fields( $field ) {
		global $wpdb;

		$field_keys = array();
		$field_keys_query = 9999;

		if( !empty( $field['value'] ) ) {
			$field_keys = json_decode( $field['value'], true );
			$field_keys_query = "'" . implode( "', '", $field_keys ) . "'";
		}

		$fields = $wpdb->get_results( "SELECT meta_key, meta_value, post_id FROM $wpdb->postmeta WHERE post_ID IN (SELECT ID FROM $wpdb->posts WHERE post_type = 'acf' AND post_status = 'publish' ) AND meta_key IN ( $field_keys_query ) " );

		$sortable_fields = array();
		foreach( $fields as $field ) {
			$sortable_fields[$field->meta_key] = $field;
		}

		$fields = array();
		foreach( $field_keys as $key ) {
			$fields[] = $sortable_fields[$key];
		}

		return $fields;

	}


}


// create field
new acf_field_field_selector();

?>
